modules/vault-cluster/main.tf

```
@@ -268,8 +268,13 @@ data "aws_iam_policy_document" "instance_role" {
   }
 }
 
+locals {
+  s3_bucket_arn = var.use_existing_s3_bucket ? element(concat(data.aws_s3_bucket.vault_storage.*.arn, [""]), 0) : element(concat(aws_s3_bucket.vault_storage.*.id, [""]), 0)
+  s3_bucket_id = var.use_existing_s3_bucket ? element(concat(data.aws_s3_bucket.vault_storage.*.id, [""]), 0) : element(concat(aws_s3_bucket.vault_storage.*.id, [""]), 0)
```

The `element(concat(..., [""]), 0)` workaround is fairly old code. Now that we're on newer versions of Terraform, we can switch to the following safer workaround:

```suggestion
  s3_bucket_arn = (
    length(aws_s3_bucket.vault_storage) > 0 
    ? aws_s3_bucket.vault_storage[0].arn 
    : length(data.aws_s3_bucket.vault_storage) > 0 ? data.aws_s3_bucket.vault_storage[0].arn : null
  )  
  s3_bucket_id = (
    length(aws_s3_bucket.vault_storage) > 0 
    ? aws_s3_bucket.vault_storage[0].id 
    : length(data.aws_s3_bucket.vault_storage) > 0 ? data.aws_s3_bucket.vault_storage[0].id : null
  )
```

---

I must admit, the new method we should use is not easier to understand.  Thanks for pointing out how I should upgrade an old habit though!